README.md
# OpenBCI Cyton BLE RFDuino Firmware

<p align="center">
  <img alt="banner" src="/images/OBCI_32bit_top.jpg/" width="600">
</p>
<p align="center" href="">
  Drive BLE communications over the RFDuino on the OpenBCI Cyton
</p>

## Welcome!

First and foremost, Welcome! :tada: Willkommen! :confetti_ball: Bienvenue! :balloon::balloon::balloon:

Thank you for visiting the OpenBCI Cyton BLE RFDuino Firmware repository.

This document (the README file) is a hub to give you some information about the project. Jump straight to one of the sections below, or just scroll down to find out more.

* [What are we doing? (And why?)](#what-are-we-doing)
* [Who are we?](#who-are-we)
* [What do we need?](#what-do-we-need)
* [How can you get involved?](#get-involved)
* [Get in touch](#contact-us)
* [Find out more](#find-out-more)
* [Understand the jargon](#glossary)

## What are we doing?

### The problem

* People have to use a dongle to get data from the Cyton
* People can't send data from the Cyton to the web browser :sad_face:
* There is a BLE switch on the Cyton that is doing nothing!

So, these problems add up to limit the amount of devices the cyton can stream it's high quality data to, and that's sad.

### The solution

The OpenBCI Cyton BLE RFDuino Firmware SDK will:

* Advertise a device id as a ble device
* Take data from the Pic32 and send it over BLE
* Prove that BLE is a stable option for the Cyton platform

Using BLE allows for every modern day computer to get data from the Cyton, so let's get this firmware stable!

## Who are we?

The author of the OpenBCI Cyton BLE RFDuino Firmware is [AJ Keller][link_aj_keller] and he was sponsored by [NEBA Health, LLC][link_neba].

We are the OpenBCI community at large though, sourcing our work from dozens of different minds for years. We love brains, we love computers, and we know a bright future is upon us then when we can stream Cyton data into our web browsers and phones.

## What do we need?

**You**! In whatever way you can help.

We need expertise in programming, user experience, software sustainability, documentation and technical writing and project management.

We'd love your feedback along the way.

Our primary goal is to drive BLE communications over the RFDuino on the OpenBCI Cyton and we're excited to support the professional development of any and all of our contributors. If you're looking to learn to code, try out working collaboratively, or translate you skills to the digital domain, we're here to help.

## Get involved

If you think you can help in any of the areas listed above (and we bet you can) or in any of the many areas that we haven't yet thought of (and here we're *sure* you can) then please check out our [contributors' guidelines](CONTRIBUTING.md) and our [roadmap](ROADMAP.md).

Please note that it's very important to us that we maintain a positive and supportive environment for everyone who wants to participate. When you join us we ask that you follow our [code of conduct](CODE_OF_CONDUCT.md) in all interactions both on and offline.

## Contact us

If you want to report a problem or suggest an enhancement we'd love for you to [open an issue](../../issues) at this github repository because then we can get right on it. But you can also contact [AJ][link_aj_keller] by email (pushtheworldllc AT gmail DOT com) or on [twitter](https://twitter.com/aj-ptw).

You can also hang out, ask questions and share stories in the [OpenBCI NodeJS room](https://gitter.im/OpenBCI/OpenBCI_NodeJS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge) on Gitter.

## Find out more

You might be interested in:

* Check out [NEBA Health, LLC][link_neba]
* An example: [RadioDevice32bit.ino][link_cyton_ble_example]

And of course, you'll want to know our:

* [Contributors' guidelines](CONTRIBUTING.md)
* [Roadmap](ROADMAP.md)

## Thank you

Thank you so much (Danke schön! Merci beaucoup!) for visiting the project and we do hope that you'll join us on this amazing journey to make programming with OpenBCI fun and easy.

# Documentation

### Table of Contents:
---

1. [Installation](#install)
2. [TL;DR](#tldr)
3. [WiFi](#wifi-docs)
  1. [General Overview](#general-overview)
  2. [Classes](#classes)
4. [Developing](#developing)
5. [Testing](#developing-testing)
6. [Contribute](#contribute)
7. [License](#license)

## <a name="developing"></a> Developing:
### <a name="developing-running"></a> Running:

```
npm install
```

### <a name="developing-testing"></a> Testing:

```
npm test
```

## <a name="contribute"></a> Contribute:

1. Fork it!
2. Branch off of `development`: `git checkout development`
2. Create your feature branch: `git checkout -b my-new-feature`
3. Make changes
4. If adding a feature, please add test coverage.
5. Ensure tests all pass. (`npm test`)
6. Commit your changes: `git commit -m 'Add some feature'`
7. Push to the branch: `git push origin my-new-feature`
8. Submit a pull request. Make sure it is based off of the `development` branch when submitting! :D

## <a name="license"></a> License:

MIT

[link_aj_keller]: https://github.com/aj-ptw
[link_shop_wifi_shield]: https://shop.openbci.com/collections/frontpage/products/wifi-shield?variant=44534009550
[link_shop_ganglion]: https://shop.openbci.com/collections/frontpage/products/pre-order-ganglion-board
[link_shop_cyton]: https://shop.openbci.com/collections/frontpage/products/cyton-biosensing-board-8-channel
[link_shop_cyton_daisy]: https://shop.openbci.com/collections/frontpage/products/cyton-daisy-biosensing-boards-16-channel
[link_ptw]: https://www.pushtheworldllc.com
[link_neba]: https://nebahealth.com
[link_openbci]: http://www.openbci.com
[link_mozwow]: http://mozillascience.github.io/working-open-workshop/index.html
[link_cyton_ble_example]: examples/RadioDevice32bit/RadioDevice32bit.ino
[link_openleaderscohort]: https://medium.com/@MozOpenLeaders
[link_mozsci]: https://science.mozilla.org


changelog.md
# v0.1.1

* Updated `bufferBLETailSend` to move the tail of the buffer before sending the packet.

# v0.1.0

New Library

# v0.0.1

Initial release


OpenBCI_Radios.h
/**
* Name: OpenBCI_Radio.h
* Date: 3/15/2016
* Purpose: This is the header file for the OpenBCI radios. Let us define two
*   over arching paradigms: Host and Device, where:
*     Host is connected to PC via USB VCP (FTDI).
*     Device is connectedd to uC (PIC32MX250F128B with UDB32-MX2-DIP).
*
* Author: Push The World LLC (AJ Keller)
*   Much credit must also go to Joel Murphy who with Conor Russomanno and Leif
*     Percifield created the original OpenBCI_32bit_Device.ino and
*     OpenBCI_32bit_Host.ino files in the Summer of 2014. Much of this code base
*     is inspired directly from their work.
*/


#ifndef __OpenBCI_Radios__
#define __OpenBCI_Radios__

#include <Arduino.h>
#include <RFduinoGZLL.h>

// needed for enum and callback support
// #include "libRFduinoGZLL.h"
#include "OpenBCI_Radios_Definitions.h"

class OpenBCI_Radios_Class {

public:
    // ENUMS
    typedef enum STREAM_STATE {
        STREAM_STATE_INIT,
        STREAM_STATE_STORING,
        STREAM_STATE_TAIL,
        STREAM_STATE_READY
    };
    typedef enum STORE_STATE {
        STORE_STATE_INIT,
        STORE_STATE_STORING_FIRST,
        STORE_STATE_STORING_SECOND,
        STORE_STATE_STORING_THRID,
        STORE_STATE_READY
    };
    // STRUCTS
    typedef struct {
        char      data[OPENBCI_MAX_PACKET_SIZE_BYTES];
        uint8_t   positionRead;
        uint8_t   positionWrite;
    } PacketBuffer;

    typedef struct {
        boolean         overflowed;
        uint8_t         numberOfPacketsToSend;
        uint8_t         numberOfPacketsSent;
        PacketBuffer    packetBuffer[OPENBCI_NUMBER_SERIAL_BUFFERS];
    } Buffer;

    typedef struct {
        uint8_t         typeByte;
        char            data[OPENBCI_MAX_PACKET_SIZE_BYTES];
        uint8_t         bytesIn;
        boolean         flushing;
        STREAM_STATE    state;
    } StreamPacketBuffer;

    typedef struct {
        uint8_t         typeByte;
        char            data[RFDUINO_BLE_MAX_PACKET_SIZE_BYTES];
        uint8_t         bytesIn;
        boolean         flushing;
        STORE_STATE     state;
    } BLESendPacketBuffer;

    typedef struct {
        boolean flushing;
        boolean gotAllPackets;
        char    data[OPENBCI_BUFFER_LENGTH_MULTI];
        int     positionWrite;
        uint8_t previousPacketNumber;
    } BufferRadio;

// SHARED
    OpenBCI_Radios_Class();
    void        begin(uint8_t);
    void        begin(uint8_t, uint32_t);
    void        beginDebug(uint8_t, uint32_t);
    void        bufferAddTimeSyncSentAck(void);
    void        bufferCleanChar(char *, int);
    void        bufferCleanCompleteBuffer(Buffer *, int);
    void        bufferCleanCompletePacketBuffer(PacketBuffer *, int);
    void        bufferCleanPacketBuffer(PacketBuffer *,int);
    void        bufferCleanBuffer(Buffer *, int);
    boolean     bufferRadioAddData(BufferRadio *, char *, int, boolean);
    void        bufferRadioClean(BufferRadio *);
    boolean     bufferRadioHasData(BufferRadio *);
    void        bufferRadioFlush(BufferRadio *);
    void        bufferRadioFlushBuffers(void);
    boolean     bufferRadioLoadingMultiPacket(BufferRadio *buf);
    byte        bufferRadioProcessPacket(char *data, int len);
    void        bufferRadioProcessSingle(BufferRadio *buf);
    boolean     bufferRadioReadyForNewPage(BufferRadio *buf);
    void        bufferRadioReset(BufferRadio *);
    boolean     bufferRadioSwitchToOtherBuffer(void);
    void        bufferResetStreamPacketBuffer(void);
    void        bufferResetBLESendPacketBuffer(void);
    boolean     bufferSerialAddChar(char);
    boolean     bufferSerialHasData(void);
    void        bufferSerialProcessCommsFailure(void);
    void        bufferSerialReset(uint8_t);
    boolean     bufferSerialTimeout(void);
    void        bufferStreamAddChar(StreamPacketBuffer *, char);
    void        bufferStreamAddChar(BLESendPacketBuffer *, char);
    boolean     bufferStreamAddData(char *);
    void        bufferStreamFlushBuffers(void);
    boolean     bufferStreamReadyForNewPacket(StreamPacketBuffer *);
    boolean     bufferStreamReadyToSendToHost(StreamPacketBuffer *buf);
    void        bufferStreamReset(void);
    void        bufferStreamReset(StreamPacketBuffer *);
    void        bufferBLESendReset(BLESendPacketBuffer *);
    boolean     bufferStreamSendToHost(StreamPacketBuffer *buf);
    void        bufferStreamStoreData(StreamPacketBuffer *, char *);
    boolean     bufferStreamTimeout(void);
    boolean     byteIdGetIsStream(uint8_t);
    int         byteIdGetPacketNumber(uint8_t);
    byte        byteIdGetStreamPacketType(uint8_t);
    char        byteIdMake(boolean, uint8_t, char *, uint8_t);
    byte        byteIdMakeStreamPacketType(uint8_t);
    boolean     commsFailureTimeout(void);
    void        configure(uint8_t,uint32_t);
    void        configureDevice(void);
    void        configureHost(void);
    void        configurePassThru(void);
    boolean     didPCSendDataToHost(void);
    boolean     didPicSendDeviceSerialData(void);
    boolean     flashNonVolatileMemory(void);
    uint32_t    getChannelNumber(void);
    uint32_t    getPollTime(void);
    boolean     hasStreamPacket(void);
    boolean     hostPacketToSend(void);
    boolean     isATailByte(uint8_t);
    void        ledFeedBackForPassThru(void);
    // void        moveStreamPacketToTempBuffer(volatile char *data);
    boolean     needToSetChannelNumber(void);
    boolean     needToSetPollTime(void);
    byte        outputGetStopByteFromByteId(char);
    void        pollHost(void);
    boolean     pollNow(void);
    boolean     packetToSend(void);
    boolean     packetsInSerialBuffer(void);
    void        pollRefresh(void);
    void        pushRadioBuffer(void);
    void        printBaudRateChangeTo(int);
    void        printChannelNumber(char);
    void        printChannelNumberVerify(void);
    void        printCommsTimeout(void);
    void        printEOT(void);
    void        printFailure(void);
    void        printMessageToDriver(uint8_t);
    void        printPollTime(char);
    void        printSuccess(void);
    void        printValidatedCommsTimeout(void);
    void        processCommsFailureSinglePacket(void);
    boolean     processDeviceRadioCharData(char *, int);
    boolean     processHostRadioCharData(device_t, char *, int);
    byte        processOutboundBuffer(PacketBuffer *);
    byte        processOutboundBufferCharDouble(char *);
    byte        processOutboundBufferCharTriple(char *);
    boolean     processOutboundBufferForTimeSync(void);
    boolean     processRadioCharDevice(char);
    boolean     processRadioCharHost(device_t, char);
    void        resetPic32(void);
    boolean     revertToDefaultPollTime(void);
    void        revertToPreviousChannelNumber(void);
    void        sendPacketToDevice(volatile device_t, boolean);
    boolean     sendPacketToHost(void);
    void        sendPollMessageToHost(void);
    void        sendRadioMessageToHost(byte);
    void        sendStreamPackets(void);
    boolean     serialWriteTimeOut(void);
    void        setByteIdForPacketBuffer(int);
    boolean     setChannelNumber(uint32_t);
    boolean     setPollTime(uint32_t);
    void        writeBufferToSerial(char *,int);

    //////////////////////
    // SHARED VARIABLES //
    //////////////////////
    // CUSTOMS
    BufferRadio bufferRadio[OPENBCI_NUMBER_RADIO_BUFFERS];
    uint8_t currentRadioBufferNum;
    BufferRadio *currentRadioBuffer;
    uint8_t streamPacketBufferHead;
    uint8_t streamPacketBufferTail;
    Buffer bufferSerial;
    PacketBuffer *currentPacketBufferSerial;
    // BOOLEANS
    boolean debugMode;
    // CHARS
    char singleCharMsg[1];
    char singlePayLoad[1];

    StreamPacketBuffer streamPacketBuffer[OPENBCI_NUMBER_STREAM_BUFFERS];
    volatile boolean sendingMultiPacket;
    volatile boolean isWaitingForNewChannelNumber;
    volatile boolean isWaitingForNewPollTime;
    volatile unsigned long timeOfLastPoll;
    unsigned long timeOfLastMultipacketSendToHost;

    boolean channelNumberSaveAttempted;
    boolean streamPacketsHaveHeads;
    volatile boolean isWaitingForNewChannelNumberConfirmation;
    volatile boolean isWaitingForNewPollTimeConfirmation;
    volatile boolean sendSerialAck;
    volatile boolean printMessageToDriverFlag;
    volatile boolean systemUp;
    volatile boolean packetInTXRadioBuffer;

    STREAM_STATE curStreamState;

    uint8_t radioMode;
    volatile uint8_t msgToPrint;
    volatile uint8_t ackCounter;

    unsigned long lastTimeHostHeardFromDevice;
    volatile unsigned long lastTimeSerialRead;


    uint32_t radioChannel;
    uint32_t previousRadioChannel;
    uint32_t pollTime;
};

// Very important, major key to success #christmas
extern OpenBCI_Radio_BLE_Class radioBLE;

#endif // OPENBCI_RADIO_H


extras/js/ble.js
const noble = require('noble');

const rfduinoServiceUuid = '2220';
const rfduinoSubscribeCharacteristicUuid = '2221';
const writeOneCharacteristicUuid = '2222';
const writeTwoCharacteristicUuid = '2223';

noble.on('stateChange', function(state) {
  if (state === 'poweredOn') {
    //
    // Once the BLE radio has been powered on, it is possible
    // to begin scanning for services. Pass an empty array to
    // scan for all services (uses more time and power).
    //
    console.log('scanning...');
    noble.startScanning([], false);
  }
  else {
    noble.stopScanning();
  }
})

let rfduinoService = null;
let rfduinoSubscribeCharacteristic = null;
let writeOneCharacteristic = null;
let writeTwoCharacteristic = null;

let connected = false;

const processBytes = (data) => {
  console.log('data', data);
};

noble.on('discover', function(peripheral) {
  // we found a peripheral, stop scanning
  noble.stopScanning();

  //
  // The advertisment data contains a name, power level (if available),
  // certain advertised service uuids, as well as manufacturer data,
  // which could be formatted as an iBeacon.
  //
  console.log('found peripheral:', peripheral.advertisement);
  //
  // Once the peripheral has been discovered, then connect to it.
  //
  peripheral.connect(function(err) {
    //
    // Once the peripheral has been connected, then discover the
    // services and characteristics of interest.
    //
    peripheral.discoverServices([rfduinoServiceUuid], function(err, services) {
      services.forEach(function(service) {
        //
        // This must be the service we were looking for.
        //
        console.log('found service:', service.uuid);

        //
        // So, discover its characteristics.
        //
        service.discoverCharacteristics([], function(err, characteristics) {

          characteristics.forEach(function(characteristic) {
            //
            // Loop through each characteristic and match them to the
            // UUIDs that we know about.
            //
            console.log('found characteristic:', characteristic.uuid);

            if (rfduinoSubscribeCharacteristicUuid == characteristic.uuid) {
              rfduinoSubscribeCharacteristic = characteristic;
            }
            else if (writeOneCharacteristicUuid == characteristic.uuid) {
              writeOneCharacteristic = characteristic;
            }
            else if (writeTwoCharacteristicUuid == characteristic.uuid) {
              writeTwoCharacteristic = characteristic;
            }
          })

          //
          // Check to see if we found all of our characteristics.
          //
          if (rfduinoSubscribeCharacteristic &&
              writeOneCharacteristic &&
              writeTwoCharacteristic) {
            //
            // We did, so bake a pizza!
            //
            rfduinoSubscribeCharacteristic.on('read', (data) => {
              // TODO: handle all the data, both streaming and not
              console.log(`${data[1]}`);
              // processBytes(data);
            });
            rfduinoSubscribeCharacteristic.notify(true);

            connected = true;

            var crust = Buffer.from("b");
            writeOneCharacteristic.write(crust, false, function(err) {
              if (err) console.error(err);
              else console.log('wrote writeOneCharacteristic');
            });

            setTimeout(() => {
              console.log("stopping stream");
              crust = Buffer.from("s");
              writeOneCharacteristic.write(crust, false, function(err) {
                if (err) console.error(err);
                else console.log('wrote writeOneCharacteristic');
              });
              setTimeout(() => {
                console.log('exit');
                process.exit(0);
              }, 100);
            }, 3000);

          }
          else {
            console.log('missing characteristics');
          }
        })
      })
    })
  })
})


extras/js/serial.js
var SerialPort = require('serialport');
var port = new SerialPort('/dev/tty.usbserial-DB00JLWR', {
  baudrate: 115200,
  autoOpen: false
});

// Switches the port into "flowing mode"
port.on('data', (data) => {
  console.log(data.toString('utf8'));
});

let count = 0;
let portOpen = false;
setInterval(() => {
  if (portOpen) {
    port.write(`pushtheworldajkeller`, function(err) {
      if (err) {
        return console.log('Error on write: ', err.message);
      }
      if (count % 20 == 0) console.log(count, 'messages written');
      count++;
    });
  }
}, 12);

// If you write before the port is opened the write will be queued
// Since there is no callback any write errors will be emitted on an error event

// Quit on any error
port.on('error', (err) => {
  console.log(err.message);
  process.exit(1);
});

port.open((err) => {
  if (err) {
    return console.log('Error opening port: ', err.message);
  }
  portOpen = true;

  console.log("port open");
  port.write('pushtheworldajkeller');

});


OpenBCI_Radios_Definitions.h
/**
* Name: OpenBCI_Radio_Definitions.h
* Date: 3/15/2016
* Purpose: This is the header file for the OpenBCI radios definitions.
*
* Author: Push The World LLC (AJ Keller)
*   Much credit must also go to Joel Murphy who with Conor Russomanno and Leif
*     Percifield created the original OpenBCI_32bit_Device.ino and
*     OpenBCI_32bit_Host.ino files in the Summer of 2014. Much of this code base
*     is inspired directly from their work.
*/

#ifndef __OpenBCI_Radio_Definitions__
#define __OpenBCI_Radio_Definitions__

// These are helpful maximums to reference nad use in the code
#define RFDUINO_BLE_MAX_PACKET_SIZE_BYTES 20

#define OPENBCI_MAX_DATA_BYTES_IN_PACKET 31
#define OPENBCI_MAX_PACKET_SIZE_BYTES 32
#define OPENBCI_MAX_PACKET_SIZE_STREAM_BYTES 33

#define OPENBCI_TIMEOUT_PACKET_NRML_uS 500 // The time to wait before determining a multipart packet is ready to be send
#define OPENBCI_TIMEOUT_PACKET_STREAM_uS 88 // Slightly longer than it takes to send a serial byte at 115200
#define OPENBCI_TIMEOUT_PACKET_POLL_MS 48 // Poll time out length for sending null packet from device to host
#define OPENBCI_TIMEOUT_COMMS_MS 270 // Comms failure time out length. Used only by Host.

// Stream byte stuff
#define OPENBCI_STREAM_BYTE_START 0xA0
#define OPENBCI_STREAM_BYTE_STOP 0xC0

// Max buffer lengths
#define OPENBCI_BUFFER_LENGTH_MULTI 528 // 16 * 33

// Number of buffers
#define OPENBCI_NUMBER_RADIO_BUFFERS 1
#define OPENBCI_NUMBER_SERIAL_BUFFERS 16
#define OPENBCI_NUMBER_STREAM_BUFFERS 25 // This should be at least one greater than poll time divided by packet interval to allow for the ack counter.

// These are the three different possible configuration modes for this library
#define OPENBCI_MODE_DEVICE 0
#define OPENBCI_MODE_HOST 1
#define OPENBCI_MODE_PASS_THRU 2

// Pins used by the Device
#define OPENBCI_PIN_DEVICE_PCG 5
// Pins used by the Host
#define OPENBCI_PIN_HOST_LED 2
#define OPENBCI_PIN_HOST_TIME 3
#define OPENBCI_PIN_HOST_RESET 6

// roles for the RFduinoGZLL
#define RFDUINOGZLL_ROLE_HOST HOST
#define RFDUINOGZLL_ROLE_DEVICE DEVICE0

// Channel limits
#define RFDUINOGZLL_CHANNEL_LIMIT_LOWER 0
#define RFDUINOGZLL_CHANNEL_LIMIT_UPPER 25

// flash memory address for RFdunioGZLL
#define RFDUINOGZLL_FLASH_MEM_ADDR 251

// Max number of packets on the TX buffer
#define RFDUINOGZLL_MAX_PACKETS_ON_TX_BUFFER 2

// Used to determine what to send after a proccess out bound buffer
#define ACTION_RADIO_SEND_NONE 0x00
#define ACTION_RADIO_SEND_NORMAL 0x01
#define ACTION_RADIO_SEND_SINGLE_CHAR 0x02

// Used to determine the result of processing a packet
#define OPENBCI_PROCESS_RADIO_FAIL_SWITCH_LAST      0x00
#define OPENBCI_PROCESS_RADIO_FAIL_SWITCH_NOT_LAST  0x01
#define OPENBCI_PROCESS_RADIO_FAIL_MISSED_LAST      0x02
#define OPENBCI_PROCESS_RADIO_FAIL_MISSED_NOT_LAST  0x03
#define OPENBCI_PROCESS_RADIO_PASS_LAST_MULTI       0x04
#define OPENBCI_PROCESS_RADIO_PASS_LAST_SINGLE      0x05
#define OPENBCI_PROCESS_RADIO_PASS_NOT_LAST_FIRST   0x06
#define OPENBCI_PROCESS_RADIO_PASS_NOT_LAST_MIDDLE  0x07
#define OPENBCI_PROCESS_RADIO_PASS_SWITCH_LAST      0x08
#define OPENBCI_PROCESS_RADIO_PASS_SWITCH_NOT_LAST  0x09

// Stream packet EOTs
#define OPENBCI_STREAM_PACKET_HEAD 0x41
#define OPENBCI_STREAM_PACKET_TAIL 0xC0

// Commands
#define OPENBCI_HOST_PRIVATE_CMD_KEY            0xF0
#define OPENBCI_HOST_CMD_CHANNEL_GET            0x00
#define OPENBCI_HOST_CMD_CHANNEL_SET            0x01
#define OPENBCI_HOST_CMD_CHANNEL_SET_OVERIDE    0x02
#define OPENBCI_HOST_CMD_POLL_TIME_GET          0x03
#define OPENBCI_HOST_CMD_POLL_TIME_SET          0x04
#define OPENBCI_HOST_CMD_BAUD_DEFAULT           0x05
#define OPENBCI_HOST_CMD_BAUD_FAST              0x06
#define OPENBCI_HOST_CMD_SYS_UP                 0x07
#define OPENBCI_HOST_CMD_TIME_PIN_HIGH          0x08
#define OPENBCI_HOST_CMD_TIME_PIN_LOW           0x09
#define OPENBCI_HOST_CMD_BAUD_HYPER             0x0A

// Raw data packet types/codes
#define OPENBCI_PACKET_TYPE_RAW_AUX      = 3; // 0011
#define OPENBCI_PACKET_TYPE_STANDARD     = 0; // 0000
#define OPENBCI_PACKET_TYPE_TIME_SYNCED  = 1; // 0001
#define OPENBCI_PACKET_TYPE_USER_DEFINED = 2; // 0010

// Possible baud rates
#define OPENBCI_BAUD_RATE_DEFAULT 115200

// Private Radio Places
#define OPENBCI_HOST_PRIVATE_POS_KEY 1
#define OPENBCI_HOST_PRIVATE_POS_CODE 2
#define OPENBCI_HOST_PRIVATE_POS_PAYLOAD 3

#endif


OpenBCI_RFDuino_BLE_Definitions.h
/**
* Name: OpenBCI_Radio_Definitions.h
* Date: 3/15/2016
* Purpose: This is the header file for the OpenBCI radios definitions.
*
* Author: Push The World LLC (AJ Keller)
*   Much credit must also go to Joel Murphy who with Conor Russomanno and Leif
*     Percifield created the original OpenBCI_32bit_Device.ino and
*     OpenBCI_32bit_Host.ino files in the Summer of 2014. Much of this code base
*     is inspired directly from their work.
*/

#ifndef __OpenBCI_RFDuino_Definitions__
#define __OpenBCI_RFDuino_Definitions__

// These are helpful maximums to reference nad use in the code
#define RFDUINO_BLE_MAX_PACKET_SIZE_BYTES 20

#define OPENBCI_MAX_DATA_BYTES_IN_PACKET 31
#define OPENBCI_MAX_PACKET_SIZE_BYTES 32
#define OPENBCI_MAX_PACKET_SIZE_STREAM_BYTES 33

#define OPENBCI_TIMEOUT_PACKET_NRML_uS 500 // The time to wait before determining a multipart packet is ready to be send
#define OPENBCI_TIMEOUT_PACKET_STREAM_uS 88 // Slightly longer than it takes to send a serial byte at 115200
#define OPENBCI_TIMEOUT_PACKET_STREAM_uS_9600 950 // Slightly longer than it takes to send a serial byte at 9600
#define OPENBCI_TIMEOUT_PACKET_POLL_MS 48 // Poll time out length for sending null packet from device to host
#define OPENBCI_TIMEOUT_COMMS_MS 270 // Comms failure time out length. Used only by Host.

#define BYTES_PER_SAMPLE 6
#define BYTES_PER_CHANNEL 3
#define POSITION_ACCEL_BYTE 19
#define BYTES_PER_ACCEL 6
#define NUM_BLE_PACKETS 10
#define BYTES_PER_BLE_PACKET 20
#define BYTES_PER_TINY_BUF 99
// Stream byte stuff
#define OPENBCI_STREAM_BYTE_START 0xA0
#define OPENBCI_STREAM_BYTE_STOP 0xC0
// 4101000000000000000000000000000000000000000000000000000000000000C04102000000000000000000000000000000000000000000000000000000000000C04103000000000000000000000000000000000000000000000000000000000000C0
// Max buffer lengths
#define OPENBCI_BUFFER_LENGTH_MULTI 100 // 16 * 33

// Number of buffers
#define OPENBCI_NUMBER_RADIO_BUFFERS 1
#define OPENBCI_NUMBER_SERIAL_BUFFERS 2

// These are the three different possible configuration modes for this library
#define OPENBCI_MODE_DEVICE 0
#define OPENBCI_MODE_HOST 1
#define OPENBCI_MODE_PASS_THRU 2

// Pins used by the Device
#define OPENBCI_PIN_DEVICE_PCG 5
// Pins used by the Host
#define OPENBCI_PIN_HOST_LED 2
#define OPENBCI_PIN_HOST_TIME 3
#define OPENBCI_PIN_HOST_RESET 6

// roles for the RFduinoGZLL
#define RFDUINOGZLL_ROLE_HOST HOST
#define RFDUINOGZLL_ROLE_DEVICE DEVICE0

// Channel limits
#define RFDUINOGZLL_CHANNEL_LIMIT_LOWER 0
#define RFDUINOGZLL_CHANNEL_LIMIT_UPPER 25

// flash memory address for RFdunioGZLL
#define RFDUINOGZLL_FLASH_MEM_ADDR 251

// Max number of packets on the TX buffer
#define RFDUINOGZLL_MAX_PACKETS_ON_TX_BUFFER 2

// Used to determine what to send after a proccess out bound buffer
#define ACTION_RADIO_SEND_NONE 0x00
#define ACTION_RADIO_SEND_NORMAL 0x01
#define ACTION_RADIO_SEND_SINGLE_CHAR 0x02

// Used to determine the result of processing a packet
#define OPENBCI_PROCESS_RADIO_FAIL_SWITCH_LAST      0x00
#define OPENBCI_PROCESS_RADIO_FAIL_SWITCH_NOT_LAST  0x01
#define OPENBCI_PROCESS_RADIO_FAIL_MISSED_LAST      0x02
#define OPENBCI_PROCESS_RADIO_FAIL_MISSED_NOT_LAST  0x03
#define OPENBCI_PROCESS_RADIO_PASS_LAST_MULTI       0x04
#define OPENBCI_PROCESS_RADIO_PASS_LAST_SINGLE      0x05
#define OPENBCI_PROCESS_RADIO_PASS_NOT_LAST_FIRST   0x06
#define OPENBCI_PROCESS_RADIO_PASS_NOT_LAST_MIDDLE  0x07
#define OPENBCI_PROCESS_RADIO_PASS_SWITCH_LAST      0x08
#define OPENBCI_PROCESS_RADIO_PASS_SWITCH_NOT_LAST  0x09

// Stream packet EOTs
#define OPENBCI_STREAM_PACKET_HEAD 0x41
#define OPENBCI_STREAM_PACKET_TAIL 0xC0

// Commands
#define OPENBCI_HOST_PRIVATE_CMD_KEY            0xF0
#define OPENBCI_HOST_CMD_CHANNEL_GET            0x00
#define OPENBCI_HOST_CMD_CHANNEL_SET            0x01
#define OPENBCI_HOST_CMD_CHANNEL_SET_OVERIDE    0x02
#define OPENBCI_HOST_CMD_POLL_TIME_GET          0x03
#define OPENBCI_HOST_CMD_POLL_TIME_SET          0x04
#define OPENBCI_HOST_CMD_BAUD_DEFAULT           0x05
#define OPENBCI_HOST_CMD_BAUD_FAST              0x06
#define OPENBCI_HOST_CMD_SYS_UP                 0x07
#define OPENBCI_HOST_CMD_TIME_PIN_HIGH          0x08
#define OPENBCI_HOST_CMD_TIME_PIN_LOW           0x09
#define OPENBCI_HOST_CMD_BAUD_HYPER             0x0A
#define OPENBCI_HOST_CMD_BAUD_BLE               0x0B

// Raw data packet types/codes
#define OPENBCI_PACKET_TYPE_RAW_AUX      = 3; // 0011
#define OPENBCI_PACKET_TYPE_STANDARD     = 0; // 0000
#define OPENBCI_PACKET_TYPE_TIME_SYNCED  = 1; // 0001
#define OPENBCI_PACKET_TYPE_USER_DEFINED = 2; // 0010

// Possible baud rates
#define OPENBCI_BAUD_RATE_DEFAULT 115200
#define OPENBCI_BAUD_RATE_BLE 9600

// Private Radio Places
#define OPENBCI_HOST_PRIVATE_POS_KEY 1
#define OPENBCI_HOST_PRIVATE_POS_CODE 2
#define OPENBCI_HOST_PRIVATE_POS_PAYLOAD 3

#endif
